feat:Support duplicate one existing Route

batman-ezio opened the pull request:
Please answer these questions before submitting a pull request

Why submit this pull request?
 - [ ] Bugfix
 - [x] New feature provided
 - [ ] Improve performance
 - [ ]  Backport patches
 - [ ] We has added two fields for upstream in APISIX.

Support duplicate one existing Route  https://github.com/apache/apisix-dashboard/issues/1487


New feature or improvement
1: duplicate button from route list
2: check duplicate name from POST route api

Describe the details and related test reports.

1:Click duplicate button from the router list, the address will be  `routes/:rid/duplicate`
![image](https://user-images.githubusercontent.com/80014017/110097045-90769d80-7dd9-11eb-9c73-52989067d690.png)


2: update the name to make sure it's no duplicated
![image](https://user-images.githubusercontent.com/80014017/110097178-b56b1080-7dd9-11eb-9509-bfe92db8eb4b.png)



3: submit data, return to the list. The new route will be there

![image](https://user-images.githubusercontent.com/80014017/110097255-c7e54a00-7dd9-11eb-9b66-7290c14f5adb.png)

nic-chen commented on api/internal/handler/route/route.go:
```diff
@@ -307,6 +307,12 @@ func generateLuaCode(script map[string]interface{}) (string, error) {
 func (h *Handler) Create(c droplet.Context) (interface{}, error) {
 	input := c.Input().(*entity.Route)
+	// check duplicate name
+	if err := h.checkDuplicateName(c, input.Name, ""); err != nil {
```
should add this check for `Update` too. Thanks.

batman-ezio replied:
ok

nic-chen commented:
need some test cases for duplicate check, thanks. @batman-ezio

batman-ezio commented:
> need some test cases for duplicate check, thanks. @batman-ezio

@nic-chen after I check duplicate name in create/update route api. some test cases failed. I think they are use the same route name to test before.  some errors throw from the route test are very strange , like this 
```
=== RUN   TestRoute_Create/create_route_success
panic: 
assert: mock: I don't know what to return because the method call was unexpected.
	Either do Mock.On("List").Return(...) first, or remove the List() call.
	This method was unexpected:
		List(store.ListInput)
		0: store.ListInput{Predicate:(func(interface {}) bool)(0x11bf5a0), Format:(func(interface {}) interface {})(nil), PageSize:0, PageNumber:0, Less:(func(interface {}, interface {}) bool)(nil)}
	at: [store_mock.go:35 route.go:537 route.go:311 route_test.go:1005] [recovered]
	panic: 
assert: mock: I don't know what to return because the method call was unexpected.
	Either do Mock.On("List").Return(...) first, or remove the List() call.
	This method was unexpected:
		List(store.ListInput)
		0: store.ListInput{Predicate:(func(interface {}) bool)(0x11bf5a0), Format:(func(interface {}) interface {})(nil), PageSize:0, PageNumber:0, Less:(func(interface {}, interface {}) bool)(nil)}
	at: [store_mock.go:35 route.go:537 route.go:311 route_test.go:1005]

``` 
I have change the route name but still got this. can you help?

nic-chen commented:
> > need some test cases for duplicate check, thanks. @batman-ezio
> 
> @nic-chen after I check duplicate name in create/update route api. some test cases failed. I think they are use the same route name to test before. some errors throw from the route test are very strange , like this
> 
> ```
> === RUN   TestRoute_Create/create_route_success
> panic: 
> assert: mock: I don't know what to return because the method call was unexpected.
> 	Either do Mock.On("List").Return(...) first, or remove the List() call.
> 	This method was unexpected:
> 		List(store.ListInput)
> 		0: store.ListInput{Predicate:(func(interface {}) bool)(0x11bf5a0), Format:(func(interface {}) interface {})(nil), PageSize:0, PageNumber:0, Less:(func(interface {}, interface {}) bool)(nil)}
> 	at: [store_mock.go:35 route.go:537 route.go:311 route_test.go:1005] [recovered]
> 	panic: 
> assert: mock: I don't know what to return because the method call was unexpected.
> 	Either do Mock.On("List").Return(...) first, or remove the List() call.
> 	This method was unexpected:
> 		List(store.ListInput)
> 		0: store.ListInput{Predicate:(func(interface {}) bool)(0x11bf5a0), Format:(func(interface {}) interface {})(nil), PageSize:0, PageNumber:0, Less:(func(interface {}, interface {}) bool)(nil)}
> 	at: [store_mock.go:35 route.go:537 route.go:311 route_test.go:1005]
> ```
> 
> I have change the route name but still got this. can you help?


hi @batman-ezio , you need to mock `routeStore.List` for  `checkDuplicateName`

you could refer to:
https://github.com/apache/apisix-dashboard/blob/2de2c72cfa8899e98045c3059d281b90cf544404/api/internal/handler/plugin_config/plugin_config_test.go#L694-L709

Jaycean commented:
@batman-ezio 
Hi, I see that in your E2E test and unit test, when creating a route, the error code returned is 400, which is usually the wrong data format or type of the route. Please check the data format or type.

batman-ezio commented:
> @batman-ezio
> Hi, I see that in your E2E test and unit test, when creating a route, the error code returned is 400, which is usually the wrong data format or type of the route. Please check the data format or type.

hi @Jaycean  I found some test case use the same route name or empty route name for e2e test,  test cases begin to bump up after my fix one by one

Jaycean commented:
> > @batman-ezio
> > Hi, I see that in your E2E test and unit test, when creating a route, the error code returned is 400, which is usually the wrong data format or type of the route. Please check the data format or type.
> 
> hi @Jaycean I found some test case use the same route name or empty route name for e2e test, test cases begin to bump up after my fix one by one
Yes, 400 is usually caused by wrong data format or type. In E2E test, a route ID r1 is usually created and passed, and the route with ID r1 will be deleted, so there is no problem with the same name.

nic-chen commented:
@batman-ezio  would you mind if we modify your code on your branch?

batman-ezio commented:
> @batman-ezio would you mind if we modify your code on your branch?

no, please
I found those test case  in `ece/route_with_*`  does not pass Name field in the body, so their route will create with a blank  name and will failed for duplicate, can we add a logic to create/update to set the name from url params if  name is not in body or blank?

nic-chen commented:
> can we add a logic in create/update API to set the name from url params if it does not exists in the request body?

oh, no, we should not do that.
we could fix in test cases which do duplicate. I think there will not be a lot, because most of test cases will delete the routes at the end.

batman-ezio commented:
> > can we add a logic in create/update API to set the name from url params if it does not exists in the request body?
> 
> oh, no, we should not do that.
> we could fix in test cases which do duplicate. I think there will not be a lot, because most of test cases will delete the routes at the end.

ok, can you help to fix the rest test cases, thanks

nic-chen commented:
> > > can we add a logic in create/update API to set the name from url params if it does not exists in the request body?
> > 
> > 
> > oh, no, we should not do that.
> > we could fix in test cases which do duplicate. I think there will not be a lot, because most of test cases will delete the routes at the end.
> 
> ok, can you help to fix the rest test cases, thanks


sure, my pleasure. will update later.

batman-ezio commented:
> > > > can we add a logic in create/update API to set the name from url params if it does not exists in the request body?
> > > 
> > > 
> > > oh, no, we should not do that.
> > > we could fix in test cases which do duplicate. I think there will not be a lot, because most of test cases will delete the routes at the end.
> > 
> > 
> > ok, can you help to fix the rest test cases, thanks
> 
> sure, my pleasure. will update later.

thanks. I have squashed all my test case commits into one commit, and rebased from master (which includes the latest frontend e2e fix), but the test case still failed

nic-chen commented:
> > > > > can we add a logic in create/update API to set the name from url params if it does not exists in the request body?
> > > > 
> > > > 
> > > > oh, no, we should not do that.
> > > > we could fix in test cases which do duplicate. I think there will not be a lot, because most of test cases will delete the routes at the end.
> > > 
> > > 
> > > ok, can you help to fix the rest test cases, thanks
> > 
> > 
> > sure, my pleasure. will update later.
> 
> thanks. I have squashed all my test case commits into one commit, and rebased from master (which includes the latest frontend e2e fix), but the test case still failed

I will try to fix backend CI this afternoon.
@LiteSun  could you please help for the FE?

LiteSun commented:
> I will try to fix backend CI this afternoon.
> @LiteSun could you please help for the FE?

ok

nic-chen commented:
backend CI had been resolved.
please help check FE CI when you have time. @LiteSun

membphis commented:
ping @LiteSun

batman-ezio commented:
after rebase from latest master, the frontend test case is still failed. 

`should import route(s) from be test files`
failed because from click multi routes.  maybe import in foreach look needs to ensure the route is deleted. add async?

batman-ezio commented:
> ping @LiteSun

I see he fire the PR for fix, https://github.com/apache/apisix-dashboard/pull/1590/files

starsz reviewed (approved):
BackEnd LGTM. The same things should be done in other module.

membphis commented:
@batman-ezio you can make a try to merge the `master` branch

LiteSun commented:
> @batman-ezio you can make a try to merge the `master` branch

Just help you sync master branch code. let's wait for CI report 😊

starsz commented:
Wait for CI.

juzhiyuan commented:
Why backend failed?

nic-chen commented:
> Why backend failed?

need some time to check for it

LiteSun commented:
ping

nic-chen commented:
@batman-ezio  please sync codes from branch master

batman-ezio commented:
@nic-chen  This PR contains two things
1: duplicate a route 
2: check duplicate name in create/update route api

#2 break many test case and take much time. 
how about I revert the code and just keep the `#1` codes here . 

I see you have a more effective  PR to do the `#2` things  here  https://github.com/apache/apisix-dashboard/pull/1606/files

and then we can move this PR forward

nic-chen commented:
> @nic-chen This PR contains two things
> 1: duplicate a route
> 2: check duplicate name in create/update route api
> 
> `#2` break many test case and take much time.
> how about I revert the code and just keep the `#1` codes here .
> 
> I see you have a more effective PR to do the `#2` things here https://github.com/apache/apisix-dashboard/pull/1606/files
> 
> and then we can move this PR forward

OK

batman-ezio commented:
CI is still not stable, I sync from master still failes

nic-chen commented:
hi @batman-ezio  
pr #1606  has been merged. please sync codes from branch master.

juzhiyuan commented on web/config/routes.ts:
```diff
@@ -39,6 +39,10 @@ const routes = [
     component: './Route/Create',
   },
+  {
+    path: '/routes/:rid/duplicate',
```
Do we need this path?

batman-ezio replied:
> hi @batman-ezio
> pr #1606 has been merged. please sync codes from branch master.

sync from master but frontend still fails

juzhiyuan replied:
cc @guoqqqi to help resolve failed test

guoqqqi replied:
Hi, I have fixed the problem with the relevant test and wait for #1608.

batman-ezio replied:
@guoqqqi  I sync from master after your PR merged, but still failed

juzhiyuan commented on web/src/pages/Route/List.tsx:
```diff
@@ -387,6 +387,9 @@ const Page: React.FC = () => {
             }}>
               {formatMessage({ id: 'component.global.view' })}
             </Button>
+            <Button type="primary" onClick={() => history.push(`/routes/${record.id}/duplicate`)}>
```
Could we fetch data first, then submit it directly by using `create` method?

batman-ezio replied:
we can't do that because the route name is unique, we need to take user to `/routes/:rid/duplicate` and fill the copied data, user needs to change the name and then get next to submit

juzhiyuan replied:
ok, got it

LiteSun commented on web/cypress/integration/route/create-edit-duplicate-delete-route.spec.js:
```diff
@@ -163,12 +163,15 @@ context('Create and Delete Route', () => {
       cy.contains('button', 'Confirm').click();
-      cy.get(this.domSelector.notification).should('contain', this.data.deleteRouteSuccess);
-    }).bind(this);
+      cy.get(domSelector.notification).should('contain', data.deleteRouteSuccess);
```
To make the test case stable, it would be better to close the notification here.

guoqqqi commented on web/cypress/integration/route/create-edit-duplicate-delete-route.spec.js:
```diff
@@ -131,12 +132,46 @@ context('Create and Delete Route', () => {
+    const { domSelector, data } = this;
+    const routeNames = [newName, duplicateNewName];
+    routeNames.forEach(function (routeName) {
+      cy.get(`${domSelector.nameSelector} .ant-input`).clear().type(routeName);
```
```suggestion
      cy.get(domSelector.name).clear().type(routeName);
```
This will fix the CI error.

nic-chen commented:
please resolve conflicts

batman-ezio commented:
> please resolve conflicts

fixed

juzhiyuan commented on web/cypress/integration/route/create-edit-duplicate-delete-route.spec.js:
```diff
@@ -153,12 +154,46 @@ context('Create and Delete Route', () => {
+  });
+
+  it('should delete the route', function () {
+    cy.visit('/routes/list');
```
@LiteSun Do we need to visit this page by clicking Sidebar Menu?

liuxiran reviewed (approved):
LGTM， thanks

and there is a UE  enhancement suggestion: 
when copying the route user in the page of first step and click [Next] button directly, they will get `duplicate route name`error, maybe leaving the route name blank would be a better choice.

juzhiyuan commented:
ok, then let me merge.

juzhiyuan commented:
thanks! @batman-ezio
